File: app/lightwait-native-application.py
#!/usr/bin/env python3

import sys
import json
import struct
import socket


# Encode a message for transmission,
# given its content.
def encodeMessage(messageContent):
    encodedContent = json.dumps(messageContent).encode('utf-8')
    encodedLength = struct.pack('@I', len(encodedContent))
    return {'length': encodedLength, 'content': encodedContent}

# Send an encoded message to stdout
def sendMessage(encodedMessage):
    sys.stdout.buffer.write(encodedMessage['length'])
    sys.stdout.buffer.write(encodedMessage['content'])
    sys.stdout.buffer.flush()

def socket_handler():
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.bind(('127.0.0.1', 33333))

    while True:
        data, addr = sock.recvfrom(1024)
        msg = encodeMessage("python 255:0:0")
        print("msg: %s" % msg)
        sendMessage(msg)
        
socket_handler()

File: app/lightwait-native-application.js
#!/usr/bin/env node
const dgram = require('dgram');

const PORT = 33333;
const HOST = '127.0.0.1';

const server = dgram.createSocket('udp4');

server.on('message', function (message, remote) {
    sendMessage(message.toString().trim())
});

server.bind(PORT, HOST);

function sendMessage(msg) {
    var buffer = Buffer.from(JSON.stringify(msg))

    var header = Buffer.alloc(4)
    header.writeUInt32LE(buffer.length, 0)

    var data = Buffer.concat([header, buffer])
    process.stdout.write(data)
}

process.on('uncaughtException', (err) => {
    sendMessage({ error: err.toString() })
})